src/vfs/virtualFileSystemWithWatch.ts

```diff
@@ -0,0 +1,865 @@
+                            // TODO: Changing from folder => file
+                        }
+                        else {
+                            // Folder update: Nothing to do.
+                            currentEntry.modifiedTime = this.now();
+                            this.invokeFsWatches(currentEntry.fullPath, "change");
+                        }
+                    }
+                }
+                else {
+                    this.ensureFileOrFolder(fileOrDirectory, options && options.ignoreWatchInvokedWithTriggerAsFileCreate);
+                }
+            }
+        }
+
+        // Output is pretty
+        writeOutputIsTTY() {
+            return true;
+        }
+
+        getNewLine() {
+            return this.newLine;
+        }
+
+        toNormalizedAbsolutePath(s: string) {
+            return getNormalizedAbsolutePath(s, this.currentDirectory);
+        }
+
+        toFullPath(s: string) {
+            return this.toPath(this.toNormalizedAbsolutePath(s));
+        }
+
+        getHostSpecificPath(s: string) {
+            if (this.windowsStyleRoot && s.startsWith(directorySeparator)) {
+                return this.windowsStyleRoot + s.substring(1);
+            }
+            return s;
+        }
+
+        now() {
+            this.time += timeIncrements;
+            return new Date(this.time);
+        }
+
+        modifyFile(filePath: string, content: string, options?: Partial<ReloadWatchInvokeOptions>) {
+            const path = this.toFullPath(filePath);
+            const currentEntry = this.fs.get(path);
+            if (!currentEntry || !isFsFile(currentEntry)) {
+                throw new Error(`file not present: ${filePath}`);
+            }
+
+            if (options && options.invokeFileDeleteCreateAsPartInsteadOfChange) {
+                this.removeFileOrFolder(currentEntry, /*isRemoveableLeafFolder*/ false);
+                this.ensureFileOrFolder({ path: filePath, content });
+            }
+            else {
+                currentEntry.content = content;
+                currentEntry.modifiedTime = this.now();
+                this.fs.get(getDirectoryPath(currentEntry.path))!.modifiedTime = this.now();
+                if (options && options.invokeDirectoryWatcherInsteadOfFileChanged) {
+                    const directoryFullPath = getDirectoryPath(currentEntry.fullPath);
+                    this.invokeFileWatcher(directoryFullPath, FileWatcherEventKind.Changed, /*useFileNameInCallback*/ true);
+                    this.invokeFsWatchesCallbacks(directoryFullPath, "rename", currentEntry.fullPath);
+                    this.invokeRecursiveFsWatches(directoryFullPath, "rename", currentEntry.fullPath);
+                }
+                else {
+                    this.invokeFileAndFsWatches(currentEntry.fullPath, FileWatcherEventKind.Changed);
+                }
+            }
+        }
+
+        renameFile(fileName: string, newFileName: string) {

```

Comment:
I think for now any method we are not using from  tsserver directly (That is anything not on fsHost and modify/delete or ensureFileOrFolder needs to be on Test instance that derives from this class )